Implement Socket.DuplicateAndClose() for Windows

antonfirsov opened the pull request:
Fixes #1760.

The implementation is based on [Framework code](https://referencesource.microsoft.com/#System/net/System/Net/Sockets/Socket.cs,170), with the following differences:
- **Breaking change:** When trying to bind a socket to a ThreadPool (thus IOCP port), if the socket has been previously bound to IOCP, an `InvalidOperationException` is thrown, instead of an undocumented `ArgumentException`.
- `SocketInformation` is not `[Serializable]` and does not contain a [snapshot of RemoteEndpoint](https://referencesource.microsoft.com/#System/net/System/Net/Sockets/SocketInformation.cs,11). Instead, the endpoint is being restored using `getsockname` in all cases.

Commit:
reimplement socket duplication, fix #1760

Commit:
additional cleanup

alnikola commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:
```diff
@@ -17,6 +17,85 @@ public partial class Socket
+                    _rightEndPoint = ep.Create(socketAddress);
+                }
+                catch
+                {
```
It looks suspicious that we silently suppress all exceptions here. What is the motivation?

antonfirsov replied:
I seems to be a bad pattern inherited from the old framework code. Would make sense to replace `IPEndPointExtensions.Create()` with a non-throwing `IPEndPointExtensions.TryCreate()`, but I did not want to apply changes of that extent in this PR.

scalablecory replied:
If this fails, the socket will be in a bad state. I'd rather we just let it throw.

That said, I don't know why this would fail as I'd expect socket duplication to not work if the address is wrong.

antonfirsov replied:
Doesn't that apply for all call sites? If we change the pattern, it's better to alter usages everywhere.
I would prefer to change it in a separate clean-up PR everywhere rather than breaking the (bad) consistency on a single call site.

scalablecory replied:
Can you point to what other code you're comparing to? UDP would function mostly fine like this but to my knowledge this would be the first API where we can have a connected socket without an endpoint set, and the other common connection-oriented APIs (send/recv) don't not set it so it would never get set.

Failing fast here would be better than failing down the line with a null endpoint, especially if it's "free" because the exception is already getting thrown.

antonfirsov replied:
The [same exception swallowing pattern](https://github.com/dotnet/runtime/blob/5b617e604cca42aac512f6be809d2a5330586400/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3141) is present in Send/Receive implementations. I don't get the purpose of those calls though, since we are not doing anything with the `remoteEp` instances being returned.

Failing fast would be indeed better here, however the handling of (undocumented & impossiblish) exceptions coming from internals looks quite ad-hoc in existing code. Some call sites just let them leak out, others prefer to swallow them. Shouldn't be there a guideline point about this?

antonfirsov replied:
Removed the try-catch, but I'd really like to have better understanding. The code was taken from Framework.

alnikola commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -49,6 +50,43 @@ public static SocketError CreateSocket(AddressFamily addressFamily, SocketType s
+
+            fixed (byte* pinnedBuffer = socketInformation.ProtocolInformation)
+            {
+                IntPtr handle = Interop.Winsock.WSASocketW(
```
Cannot it return SafeSocketHandle directly from the Interop call? AFAIK, in other places we do this to lower chances of resource leaking.

antonfirsov replied:
Actually, WinSock P/Invoke API-s are all returning an `IntPtr` for a reason I don't know. I'd prefer to keep it for now, but worth a deeper look to have a better understanding.

alnikola commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:
```diff
@@ -0,0 +1,306 @@
+            Assert.Equal(blocking, clone.Blocking);
+        }
+
+        public class NotSupportedOnUnix
```
It's better not to nest one test class into another. It impairs readability.

antonfirsov replied:
A [hierarchical pattern](https://haacked.com/archive/2012/01/02/structuring-unit-tests.aspx/) seems actually more readable for me in many cases: helps avoiding long test names, organizing stuff into groups instead. However in this particular case it would be definitely better to ungroup them.

I wonder if we have a written guideline for test code?

scalablecory replied:
I don't think I've seen this pattern elsewhere in our tests.

Our unit testing is modeled after a book... I believe The Art of Unit Testing. @davidsh can confirm.

Commit:
ungroup PNSE tests, no RemoteExecutor timeout

scalablecory commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    internal static partial class Winsock
+    {
+        [StructLayout(LayoutKind.Sequential, CharSet=CharSet.Auto)]
```
I think we can get rid of this attribute; sequential is the default and there are no characters to marshal here.

antonfirsov replied:
Isn't `LayoutKind.Auto` the default? All pinvoke code I've ever seen keeps adding `[StructLayout(LayoutKind.Sequential)]`, I thought this ancient rite is being done for a reason 😄

`CharSet` is indeed meaningless.

scalablecory replied:
Structs use sequential layout by default in c#. It's probably being added just to make explicit that it will be used for pinvoke -- fine to keep for consistency.

scalablecory commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,52 @@
+
+        [DllImport(Interop.Libraries.Ws2_32, SetLastError = true)]
+        internal static extern unsafe SocketError WSADuplicateSocket(
+            [In] SafeHandle socketHandle,
```
More appropriate to use `SafeSocketHandle` here?

scalablecory commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,52 @@
+        internal static extern unsafe SocketError WSADuplicateSocket(
+            [In] SafeHandle socketHandle,
+            [In] uint targetProcessId,
+            [In] byte* pinnedBuffer
```
```suggestion
            [In] byte* lpProtocolInfo
```

scalablecory commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketInformation.cs:
```diff
@@ -8,5 +8,16 @@ public struct SocketInformation
+
+        internal bool GetOption(SocketInformationOptions option)
+        {
+            return ((Options & option) != 0);
```
nit: superfluous parens

scalablecory commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        [DllImport(Interop.Libraries.Ws2_32, SetLastError = true)]
+        internal static extern unsafe SocketError WSADuplicateSocket(
```
`WSADuplicateSocket` does not return a `SocketError`. It returns != 0 for error.

and then throw a new `SocketException` with default ctor or call `Marshal.GetLastWin32Error()` if you need the error.

scalablecory commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:
```diff
@@ -0,0 +1,306 @@
+            _ = client.ConnectAsync(listener1.LocalEndPoint);
+
+            using Socket handler = await listener1.AcceptAsync();
+            await client.SendAsync(TestBytes, SocketFlags.None);
```
You should wait for the `ConnectAsync` to complete before sending.

Commit:
address review findings

antonfirsov commented:
I think I addressed most review findings, except the ones dealing with existing patterns (`WSASocket` return value and the dumb catch blocks).

Should I sneak fixes for those into this PR, or shall we do it separately instead?

ManickaP commented on src/libraries/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -197,6 +197,9 @@
     <Compile Include="$(CommonPath)Interop\Windows\WinSock\Interop.WSAConnect.cs">
       <Link>Common\Interop\Windows\WinSock\Interop.WSAConnect.cs</Link>
     </Compile>
+    <Compile Include="..\..\Common\src\Interop\Windows\WinSock\Interop.WSADuplicateSocket.cs">
```
Shouldn't here be something like $(CommonPath) instead of relative path?

antonfirsov replied:
Good catch!

Commit:
fix path for Interop.WSADuplicateSocket.cs

Commit:
remove swallowing in Socket(SocketInformation)

scalablecory commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:
```diff
@@ -17,6 +17,79 @@ public partial class Socket
+            info.SetOption(SocketInformationOptions.NonBlocking, !Blocking);
+            info.SetOption(SocketInformationOptions.Listening, _isListening);
+
+            Close(-1);
```
name the param please, it's not clear what `-1` is.

scalablecory commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketInformation.cs:
```diff
@@ -8,5 +8,16 @@ public struct SocketInformation
+
+        internal bool GetOption(SocketInformationOptions option)
+        {
+            return (Options & option) != 0;
```
I would use `== option` here. It will be safer if `option` uses multiple bits.

The JIT is smart here and will output an optimal `TEST` instruction (equiv to `!= 0`) if the constant is a single bit.

scalablecory reviewed (approved):
a couple things, otherwise looks good.

Commit:
Merge branch 'master' into af/duplicate-socket

Commit:
review suggestions

antonfirsov commented:
/azp run runtime-libraries outerloop

jkotalik commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
nit: copyright

jkotalik commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -1246,5 +1284,19 @@ internal static SocketError Disconnect(Socket socket, SafeSocketHandle handle, b
+        {
+            socketInformation = new SocketInformation
+            {
+                ProtocolInformation = new byte[Interop.Winsock.WSAProtocolInfo.Size]
```
nit: stackalloc instead of newing up this array

antonfirsov replied:
This is an [existing public API](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketinformation.protocolinformation), that uses `byte[]`.

jkotalik commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -49,6 +50,43 @@ public static SocketError CreateSocket(AddressFamily addressFamily, SocketType s
+                    (AddressFamily)(-1),
+                    (SocketType)(-1),
+                    (ProtocolType)(-1),
+                    (IntPtr)pinnedBuffer, 0, Interop.Winsock.SocketConstructorFlags.WSA_FLAG_OVERLAPPED);
```
I'd add a comment why WSA_FLAG_NO_HANDLE_INHERIT isn't set here.

antonfirsov replied:
This is a good finding.

I'll be honest, this wasn't an intentional decision, I reused Framework code, where sockets were inheritable by default. I guess it's not possible to create inheritable duplicates of a non-inheritable handles but I will double-check.

jkotalik replied:
That was my assumption as well. I just noticed the discrepancy between this and the other calls to Interop.Winsock.WSASocketW across the repo: https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/Common/src/System/Net/SocketProtocolSupportPal.Windows.cs and https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

antonfirsov replied:
I would now prefer to add `WSA_FLAG_NO_HANDLE_INHERIT` here for consistency, even if it has no effect.

jkotalik replied:
From what the docs say, WSA_FLAG_NO_HANDLE_INHERIT is used to make it so the socket isn't inheritable. My assumption was that if it was set, DuplicateSocket would fail. I'd try and confirm that.

antonfirsov replied:
@jkotalik my understanding is that inheritance and duplication are different concepts. It's possible to `DuplicateHandle()` stuff into arbitrary unrelated processes, even if the handle is not inheritable, while handle inheritance is carried out in an automatic manner between parent and child processes.

Just made an experiment with sockets and it worked. However I definitely need to do some additional research to make sure my understanding is correct.

scalablecory replied:
Inheritable sockets allow child processes to access the socket; duplicate sockets allow any (non-child) process to take ownership of the socket.

scalablecory replied:
It's a good call out though; if the new default is to make all sockets non-inheritable (maybe not portable, perhaps some perf gain) then we should confirm if that's appropriate for these duplicated sockets too.

jkotalik reviewed (commented):
Have we been able to pass these bits off to internal teams to verify?

Commit:
make sure duplicate socket is not inheritable

antonfirsov commented:
Pushed a change to deal with handle inheritance. The solution is not trivial because of the esoteric socket constructor behavior. I had to use `SetHandleInformation`. (See comments.)

@jkotalik thanks for your engagement here, it prevented a bug!

Commit:
remove debug code leftover

antonfirsov commented:
/azp run runtime-libraries outerloop

antonfirsov commented:
/azp run runtime-libraries outerloop

scalablecory commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -78,7 +91,9 @@ public static SocketError CreateSocket(AddressFamily addressFamily, SocketType s
                 }
 
-                Interop.Winsock.WSAProtocolInfo protocolInfo = Marshal.PtrToStructure<Interop.Winsock.WSAProtocolInfo>((IntPtr)pinnedBuffer);
+                Interop.Kernel32.SetHandleInformation(socket, Interop.Kernel32.HandleFlags.Inherit, 0);
```
`SetHandleInformation` returns false on failure.  We should either check for failure and throw, or, if this truly has no effect (per above comment), add a comment here noting why it is not being checked.

scalablecory commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -41,14 +42,63 @@ public static SocketError GetLastSocketError()
+                if (socket.IsInvalid)
+                {
+                    return GetLastSocketError();
+                }
```
Is it possible for `NetEventSource.Info` to call some win32 method that would reset the last error before the call to `GetLastSocketError`?

stephentoub replied:
Yes, it almost certainly will.  And the socket.Dispose that should be added.

antonfirsov replied:
In that case, the other `CreateSocket` overload is also incorrect.

scalablecory reviewed (approved):
looks good after these two comments.

stephentoub commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.HandleInformation.cs:
```diff
@@ -0,0 +1,24 @@
+        {
+            None = 0,
+            Inherit = 1,
+            ProtectFromClose = 2
```
We generally try to use the same naming as in the native signatures, e.g.
HANDLE_FLAG_INHERIT
HANDLE_FLAG_PROTECT_FROM_CLOSE
etc.

antonfirsov replied:
Looks like half of the WSA interop code does not follow the guideline. I've used this naming for consistency with those parts (eg. `AsyncEventBits`). Maybe worth to fix those.

stephentoub replied:
Ok.  Let's use the desired naming, and separately fix the interop code that's not following it.

stephentoub commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.HandleInformation.cs:
```diff
@@ -0,0 +1,24 @@
+    internal static partial class Kernel32
+    {
+        [Flags]
+        internal enum HandleFlags: uint
```
```suggestion
        internal enum HandleFlags : uint
```

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+    internal static partial class Winsock
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct WSAProtocolChain
```
WSAPROTOCOLCHAIN

antonfirsov replied:
My previous statement is mostly true for structures.

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            internal int ChainLen;
+            [MarshalAs(UnmanagedType.ByValArray, SizeConst=7)]
+            internal uint[] ChainEntries;
```
Rather than allocating an actual array, can you just have this directly mirror the native definition by using a fixed buffer?
```C#
private const int MAX_PROTOCOL_CHAIN = ...;
...
internal fixed uint ChainEntries[MAX_PROTOCOL_CHAIN];
```

antonfirsov replied:
I did not bother turning stuff blittable (just reused FW code), since socket duplication doesn't look like something happening on hot application path. Will fix it.

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+            internal uint[] ChainEntries;
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet=CharSet.Auto)]
```
Did you intend for this to be Auto rather than Unicode?  In general in .NET Core we've centralized on the wide variants, e.g. WSAPROTOCOL_INFOW.

antonfirsov replied:
It shall be Unicode.

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet=CharSet.Auto)]
+        internal struct WSAProtocolInfo
```
WSAPROTOCOL_INFOW

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+        [StructLayout(LayoutKind.Sequential, CharSet=CharSet.Auto)]
+        internal struct WSAProtocolInfo
+        {
+            internal uint ServiceFlags1;
```
All of these should map to the native naming, e.g. dwServiceFlags1, etc.

I'll stop commenting on naming.

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+            internal uint MessageSize;
+            internal uint ProviderReserved;
+            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 256)]
+            internal string ProtocolName;
```
Similar comment here about using a fixed char[] buffer rather than MarshalAs and a string, especially since we don't seem to actually use this anywhere.

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 256)]
+            internal string ProtocolName;
+
+            public static readonly int Size = Marshal.SizeOf(typeof(WSAProtocolInfo));
```
Can we use sizeof instead (once you remove the MarshalAs)?

stephentoub commented on src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSADuplicateSocket.cs:
```diff
@@ -0,0 +1,56 @@
+        internal static extern unsafe int WSADuplicateSocket(
+            [In] SafeSocketHandle socketHandle,
+            [In] uint targetProcessId,
+            [In] byte* lpProtocolInfo
```
Why `byte*` rather than `WSAPROTOCOL_INFOW*`?  Once you remove the MarshalAs and this becomes a blittable type, can that be changed?

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs:
```diff
@@ -59,6 +59,7 @@ internal ThreadPoolBoundHandle GetOrAllocateThreadPoolBoundHandle(bool trySkipCo
                     catch (Exception exception) when (!ExceptionCheck.IsFatal(exception))
                     {
                         bool closed = IsClosed;
+                        bool alreadyBound = !IsInvalid && !IsClosed && (exception is ArgumentException);
```
Won't we get an ArgumentException as well if the handle wasn't opened for overlapped I/O?  We're assuming that's the case then?

antonfirsov replied:
I made this based this on my understanding of the code in `ClrThreadPoolBoundHandle.Windows.cs`:

https://github.com/dotnet/runtime/blob/6ed4a5e3db0e0cefb6924881a99888a6e192a0f2/src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs#L30-L34

I assume that the handle can not be invalid since we checked it before. Based on the comment, and the exception message, the only other possible reason is "Handle already bound or sync handle".  Let me know, if I'm missing something!
(The docs for win32 `BindIoCompletionCallback`  were not really helpful to me.)

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs:
```diff
@@ -67,6 +68,12 @@ internal ThreadPoolBoundHandle GetOrAllocateThreadPoolBoundHandle(bool trySkipCo
+
+                        if (alreadyBound)
+                        {
+                            throw new InvalidOperationException(SR.net_sockets_asyncoperations_notallowed);
```
Worth including `exception` as an inner exception here?

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs:
```diff
@@ -11,6 +11,25 @@ namespace System.Net.Sockets
+        public SocketInformation DuplicateAndClose(int targetProcessId)
+        {
+            //
+            // DuplicateAndClose is not supported on Unix, since passing FD-s between processes
```
```suggestion
            // DuplicateAndClose is not supported on Unix, since passing file descriptors between processes
```

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs:
```diff
@@ -11,6 +11,25 @@ namespace System.Net.Sockets
+        {
+            //
+            // DuplicateAndClose is not supported on Unix, since passing FD-s between processes
+            // should involve Unix Domain Sockets. The programming model is fundamentally different,
```
```suggestion
            // requires Unix Domain Sockets. The programming model is fundamentally different,
```

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs:
```diff
@@ -11,6 +11,25 @@ namespace System.Net.Sockets
+            //
+            // DuplicateAndClose is not supported on Unix, since passing FD-s between processes
+            // should involve Unix Domain Sockets. The programming model is fundamentally different,
+            // and incompatible with the design of SocketInformation API-s.
```
```suggestion
            // and incompatible with the design of SocketInformation-related methods.
```

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs:
```diff
@@ -11,6 +11,25 @@ namespace System.Net.Sockets
+            //
+            // This constructor works in conjunction with DuplicateAndClose, which is not supported on Unix.
+            // See comments in DuplicateAndClose.
+            //
```
Do we use this style of commenting elsewhere in System.Net.Sockets, with a blank comment line above and below?  We've generally moved away from that elsewhere.

antonfirsov replied:
Occasionally yes, it was was used in the methods I changed. Getting rid of it.

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:
```diff
@@ -17,6 +17,79 @@ public partial class Socket
+            }
+
+            if (_handle.IsInvalid)
+            {
```
We should dispose of state here that may have been created, namely the _handle.  Even if it's invalid, it'll likely still trigger finalization.

antonfirsov replied:
This looks like an unnecessary check, since `SocketPal.CreateSocket` already validates handle. Added the disposal there, getting rid of this branch here.

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:
```diff
@@ -17,6 +17,79 @@ public partial class Socket
+            }
+
+            if (_addressFamily != AddressFamily.InterNetwork && _addressFamily != AddressFamily.InterNetworkV6)
+            {
```
And definitely here we should dispose _handle, since we actually have a valid handle at this point.

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -41,14 +42,63 @@ public static SocketError GetLastSocketError()
+            {
+                // Sockets are non-inheritable in .NET Core.
+                // Handle properties like HANDLE_FLAG_INHERIT are not cloned with socket duplication, therefore
+                // we need to disable handle inheritance when constructing the new socket handle from Protcol Info.
```
```suggestion
                // we need to disable handle inheritance when constructing the new socket handle from Protocol Info.
```

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -41,14 +42,63 @@ public static SocketError GetLastSocketError()
+                if (NetEventSource.IsEnabled) NetEventSource.Info(null, socket);
+
+                if (socket.IsInvalid)
+                {
```
socket.Dispose();

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:
```diff
@@ -41,14 +42,63 @@ public static SocketError GetLastSocketError()
+
+                Interop.Kernel32.SetHandleInformation(socket, Interop.Kernel32.HandleFlags.Inherit, 0);
+
+                Interop.Winsock.WSAProtocolInfo protocolInfo = Marshal.PtrToStructure<Interop.Winsock.WSAProtocolInfo>((IntPtr)pinnedProtocolInformation);
```
We can avoid using PtrToStructure once the previously discussed changes are made around making the struct blittable.

Commit:
Merge remote-tracking branch 'origin/master' into af/duplicate-socket

Commit:
harden SocketPal.CreateSocket()

Commit:
Merge branch 'master' into af/duplicate-socket

Commit:
blittable WSAPROTOCOL_INFOW

Commit:
additional naming fixes,

default System.Net.Sockets.sln configurations to Windows_NT again

Commit:
disposal on errors, improve comments, more tests

stephentoub commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:
```diff
@@ -17,6 +17,77 @@ public partial class Socket
+
+            Internals.SocketAddress socketAddress = IPEndPointExtensions.Serialize(ep);
+            errorCode = SocketPal.GetSockName(_handle, socketAddress.Buffer, ref socketAddress.InternalSize);
+            if (errorCode == SocketError.Success)
```
And if it's not success, it's ok to do nothing?

antonfirsov replied:
Framework does not throw when this fails, I left it like that for compatibility. Was it a bad approach? Is it better to be in favor of fail-fast instead?

scalablecory replied:
My preference would be to not have a connected socket constructed without a `_rightEndPoint`, as it would not get set elsewhere.

I'm not sure why `getsockname` would return an error here other than for an unbound socket. I feel like this could reasonably be:

```c#
if(errorCode == SocketError.Success)
{
   _rightEndPoint = ep.Create(socketAddress);
}
else if(errorCode == SocketError.InvalidArgument)
{
   // Socket is not yet bound.
}
else
{
   throw new SocketException((int)errorCode);
}
```

antonfirsov replied:
What do we do when an "if this happens it's definitely a bug in our code" execution path gets hit?
Shall we throw? Or just `Debug.Assert()` and forget?

scalablecory replied:
I'd say assert if we are checking against our own code, throw if we are checking against external code.

stephentoub commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:
```diff
@@ -0,0 +1,391 @@
+    public class SocketDuplicationTests
+    {
+        private readonly ArraySegment<byte> _receiveBuffer = new ArraySegment<byte>(new byte[32]);
+        const string TestMessage = "test123!";
```
```suggestion
        private const string TestMessage = "test123!";
```

stephentoub commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:
```diff
@@ -0,0 +1,391 @@
+        private static ArraySegment<byte> TestBytes => Encoding.ASCII.GetBytes(TestMessage);
+        private static string GetMessageString(ArraySegment<byte> data, int count) =>
+            Encoding.ASCII.GetString(data.AsSpan().Slice(0, count));
+        private static int CurrentProcessId => Process.GetCurrentProcess().Id;
```
Nit: ideally this would dispose of the Process returned from GetCurrentProcess

stephentoub commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:
```diff
@@ -0,0 +1,391 @@
+            Assert.Equal(TestMessage, receivedMessage);
+        }
+
+        [OuterLoop] // long-running
```
What makes it long-running?

antonfirsov replied:
For the same reason `CreateSocket.CtorAndAccept_SocketNotKeptAliveViaInheritance` is running for about 2 sec in debug on my PC. There are nested child processes + pipes. I'm also wondering why is it that much.

stephentoub replied:
We should figure out why it's taking 2 seconds.  It shouldn't be that long.

antonfirsov replied:
Since this also concerns other tests, I'd rather take care of it after separately. I'm preparing a test fix PR anyways.

antonfirsov commented on src/libraries/Common/tests/System/Net/Sockets/SocketTestExtensions.cs:
```diff
@@ -36,5 +36,18 @@ public static void ForceNonBlocking(this Socket socket, bool force)
+
+        public static (Socket, Socket) CreateConnectedSocketPair()
+        {
+            using Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
```
To me it looks quite arbitrary whether a functionality is also tested against IPV6 in socket test code. I added a few IPV6 test cases for  the duplication to make sure it is covered. See the parameters of `DuplicateAndClose_TcpServerHandler` which is emulating most common user scenario.

antonfirsov commented on src/libraries/System.Net.Sockets/System.Net.Sockets.sln:
```diff
@@ -30,8 +30,8 @@ Global
 		{8CBA022C-635F-4C8D-9D29-CD8AAC68C8E6}.Release|Any CPU.Build.0 = netcoreapp5.0-Windows_NT-Release|Any CPU
-		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.Debug|Any CPU.ActiveCfg = netcoreapp5.0-Unix-Debug|Any CPU
-		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.Debug|Any CPU.Build.0 = netcoreapp5.0-Unix-Debug|Any CPU
+		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.Debug|Any CPU.ActiveCfg = netcoreapp5.0-Windows_NT-Debug|Any CPU
```
@stephentoub all other solutions default to Windows_NT. I suppose you altered this by accident in #2063.

stephentoub replied:
If I modified it, it was VS modifying it automatically.  I'd need to look at the various guids, but there is a unix build in this project, so it would make sense for there to be unix entries here.

Commit:
nits

Commit:
handle GetSockName() error

antonfirsov commented:
/azp run runtime-libraries outerloop
